src/kaboom.ts

```diff
@@ -5563,9 +5908,110 @@ export default (gopt: KaboomOpt = {}): KaboomCtx => {
 				return numRows * opt.tileHeight
 			},
 
+			tile2Pos(tilePos: Vec2) {
+				return tilePos.scale(opt.tileWidth, opt.tileHeight)
+			},
+	
+			pos2Tile(pos: Vec2) {
+				return vec2(Math.floor(pos.x / opt.tileWidth), Math.floor(pos.y / opt.tileHeight))
+			},
+	
+			getSpatialMap() {
+				if (!spatialMap) {
+					this._createSpatialMap()
+				}
+				return spatialMap
+			},
+	
+			onSpatialMapChanged(this: GameObj<LevelComp>, cb: () => void) {
+				return this.on("spatial_map_changed", cb)
+			},
+	
+			onNavigationMapInvalid(this: GameObj<LevelComp>, cb: () => void) {
+				return this.on("navigation_map_invalid", cb)
+			},
+	
+			getAt(tilePos: Vec2) {
+				if (!spatialMap) {
+					this._createSpatialMap()
+				}
+				const hash = this._tile2Hash(tilePos)
+				return spatialMap[hash] || []
+			},
+	
+			update() {
+				if (spatialMap) {
+					this._updateSpatialMap()
+				}
+			},
+
+			/**
+			 * Private methods start here
+			 */
+
+			_tile2Hash(tilePos: Vec2) {
+				return tilePos.x + tilePos.y * numColumns
+			},
+	
+			_hash2Tile(hash: number) {
+				return vec2(Math.floor(hash % numColumns), Math.floor(hash / numColumns))
+			},
+	
+			// The spatial map keeps track of the objects at each location
+			_createSpatialMap(this: GameObj<LevelComp>) {
+				spatialMap = []
+				for (const child of this.children) {
+					this._insertIntoSpatialMap(child)
+				}
+			},
+	
+			_updateSpatialMap(this: GameObj<LevelComp>) {
+				let spatialMapChanged = false
+				let navigationMapChanged = false
+				for (const child of this.children) {
+					const tilePos = this.pos2Tile(child.pos)
+					if (child.tilePos.x != tilePos.x || child.tilePos.y != tilePos.y) {
+						spatialMapChanged = true
+						navigationMapChanged = navigationMapChanged || (child.isObstacle !== undefined || child.cost)
+						this._removeFromSpatialMap(child)
+						child.tilePos.x = tilePos.x
+						child.tilePos.y = tilePos.y
+						this._insertIntoSpatialMap(child)
+					}
+				}
+				if (spatialMapChanged) {
+					this.trigger("spatial_map_changed")
+				}
+				if (navigationMapChanged) {
+					this.trigger("navigation_map_invalid")
+				}
+			},
+	
+			_insertIntoSpatialMap(object: GameObj<any>) {
+				const hash = this._tile2Hash(object.tilePos)
+				if (spatialMap[hash]) {
+					spatialMap[hash].push(object)
+				}
+				else {
+					spatialMap[hash] = [object]
+				}
+			},
+	
+			_removeFromSpatialMap(object: GameObj<any>) {
+				const hash = this._tile2Hash(object.tilePos)
+				if (spatialMap[hash]) {
+					const index = spatialMap[hash].indexOf(object)
+					if (index >= 0) {
+						spatialMap[hash].splice(index, 1)
+					}
+				}
+			},
 		}
 
 		level.use(levelComp)
+		if (opt.usePathfinding) {
+			level.use(pathfinding())
+		}
```

Comment:
I'm thinking if we can omit this option, and use pathfinding if user call any of the pathfinding functions, so from the UX side user have one less option to care about and can always assume addLevel has pathfinding capabilities

Reply:
We could test this in getPath/getTilePath and add the component if needed yes.